middlewares/ip/ip_whitelist.go

```diff
@@ -0,0 +1,43 @@
+// Copyright 2016 The Chihaya Authors. All rights reserved.
+// Use of this source code is governed by the BSD 2-Clause license,
+// which can be found in the LICENSE file.
+
+package ip
+
+import (
+	"net"
+
+	"github.com/chihaya/chihaya"
+	"github.com/chihaya/chihaya/config"
+	"github.com/chihaya/chihaya/server/store"
+	"github.com/chihaya/chihaya/tracker"
+)
+
+func init() {
+	tracker.RegisterAnnounceMiddleware("ip_whitelist", whitelistAnnounceIP)
+}
+
+// whitelistAnnounceIP provides a middleware that only allows IPs to announce
+// that are stored in an IPStore.
+func whitelistAnnounceIP(next tracker.AnnounceHandler) tracker.AnnounceHandler {
+	return func(cfg *config.TrackerConfig, req *chihaya.AnnounceRequest, resp *chihaya.AnnounceResponse) (err error) {
+		var whitelisted bool
+
+		// We have to check explicitly if they are present, because someone
+		// could have added a <nil> net.IP to the store.
+		if req.IPv4 != nil && req.IPv6 != nil {
+			whitelisted, err = store.MustGetStore().HasAllIPs([]net.IP{req.IPv4, req.IPv6})
```

Comment:
This function just returns the inner-function, though. It doesn't get called until a request flows through the tracker. I change I suggested was to remove the verbosity of calling MustGetStore() over and over again, it shouldn't change the semantics of anything.